core/src/main/java/org/elasticsearch/index/translog/MultiSnapshot.java

```diff
@@ -84,75 +84,41 @@ public void close() throws IOException {
         onClose.close();
     }
 
-    /**
-     * A wrapper of {@link FixedBitSet} but allows to check if all bits are set in O(1).
-     */
-    private static final class CountedBitSet {
-        private short onBits;
-        private final FixedBitSet bitset;
-
-        CountedBitSet(short numBits) {
-            assert numBits > 0;
-            this.onBits = 0;
-            this.bitset = new FixedBitSet(numBits);
-        }
-
-        boolean getAndSet(int index) {
-            assert index >= 0;
-            boolean wasOn = bitset.getAndSet(index);
-            if (wasOn == false) {
-                onBits++;
-            }
-            return wasOn;
-        }
-
-        boolean hasAllBitsOn() {
-            return onBits == bitset.length();
-        }
-    }
-
-    /**
-     * Sequence numbers from translog are likely to form contiguous ranges,
-     * thus collapsing a completed bitset into a single entry will reduce memory usage.
-     */
     static final class SeqNoSet {
         static final short BIT_SET_SIZE = 1024;
-        private final LongSet completedSets = new LongHashSet();
-        private final LongObjectHashMap<CountedBitSet> ongoingSets = new LongObjectHashMap<>();
+        private final LongObjectHashMap<BitSet> bitSets = new LongObjectHashMap<>();
 
         /**
          * Marks this sequence number and returns <tt>true</tt> if it is seen before.
          */
         boolean getAndSet(long value) {
             assert value >= 0;
             final long key = value / BIT_SET_SIZE;
-
-            if (completedSets.contains(key)) {
-                return true;
-            }
-
-            CountedBitSet bitset = ongoingSets.get(key);
+            BitSet bitset = bitSets.get(key);
             if (bitset == null) {
                 bitset = new CountedBitSet(BIT_SET_SIZE);
-                ongoingSets.put(key, bitset);
-            }
-
-            final boolean wasOn = bitset.getAndSet(Math.toIntExact(value % BIT_SET_SIZE));
-            if (bitset.hasAllBitsOn()) {
-                ongoingSets.remove(key);
-                completedSets.add(key);
+                bitSets.put(key, bitset);
             }
+            final int index = Math.toIntExact(value % BIT_SET_SIZE);
+            final boolean wasOn = bitset.get(index);
+            bitset.set(index);
             return wasOn;
         }
 
         // For testing
         long completeSetsSize() {
```

Comment:
Now that we abstract the underlying storage away, I don't think we care about whether underlying sets are completed or not. All we care about is functionality. I think we can remove completedSetSize and ongoingSetSize and their tests?

Reply:
I pushed https://github.com/elastic/elasticsearch/pull/27547/commits/14aca62e03990e71fc98bbc4b97e9bca9d7d9567